src/DnsClient/DnsMessageHandler.cs

```diff
@@ -48,6 +48,14 @@ internal abstract class DnsMessageHandler
             return false;
         }
 
+        protected static void ValidateResponse(DnsRequestMessage request, DnsResponseMessage response)
+        {
+            if (request.Header.Id != response.Header.Id)
```

Comment:
right now, response can be potentially be null at this point, at least if the TCP handler is used.
I'll change that in another PR.
But maybe better do a null check and return right away, if response or request are null.